### ISSUE TITLE
Bug with `getReferenceLinkImageData` and Brackets

### ISSUE THREAD

**Issue by dsanders11** [DavidAnson/markdownlint#684]
Seems like `getReferenceLinkImageData` from `markdownlint/helpers` struggles with links which have brackets inside of the link text, such as:

* `[foo[]](https://example.com/)`
* `[foo[bar]](https://example.com/)`
* `[foo[](https://example.com/)`
* `[foo[]][]`

When I get `shortcuts` from the result of `getReferenceLinkImageData`, it will include shortcuts like `foo[`, even though none of those are shortcut links.

It's unfortunately a shortcoming of the regexes from what I saw, so maybe 8d6e0b541958e044c73a12489d5e8dcf29467a3d will help. Those are valid links, they're just tricky to parse with regexes since there are extra brackets in there. Balanced brackets are allowed in the link text, but good luck writing a regex which can differentiate balanced brackets. 😅 Opening brackets are also valid, but trip up `getReferenceLinkImageData` as well.

**Comment by DavidAnson** [comment]
@dsanders11, could you please have a look at <https://github.com/DavidAnson/markdownlint/commit/f000a33481fb1401ca0052dd93bbaedaddea34ec>? I checked your examples above and don't see anything wrong with the current behavior in `next` branch. (Is it possible one of my other pre-release changes fixed your scenario?)

Specifically, I added a test to make sure the `shortcuts` `Set` is empty for the following input:

```markdown
Full reference link: [text0][label]
Collapsed reference link: [label][]
Nested empty brackets: [text1[]](https://example.com/)
Missing close bracket, empty text: [text2[](https://example.com/)
Empty bracket pairs: [text3[]][]

[label]: https://example.com/label
```

That covers your first, third, and fourth examples. Your second example is omitted because `bar` is a valid shortcut link if `[bar]` is defined (and `shortcuts` doesn't try to verify that by design):

```markdown
[foo[bar]](https://example.com/)

[bar]: https://example.com/
```

If you have any other examples of wrong-seeming behavior, please share them here.

**Comment by DavidAnson** [comment]
LOL, yes, cherry-picking the new test commit above to `main` branch and running it produces errors just as you said: `text1[, text3[`. Good, something else I already did made this better for the next release! :)

**Comment by DavidAnson** [comment]
The fix was most likely this commit: <https://github.com/DavidAnson/markdownlint/commit/6974cd6012afbcb3bce95ff78d177a1730845202>.

**Comment by dsanders11** [comment]
Thanks @DavidAnson!
